proposals/NNNN-concurrency-objc.md

```diff
@@ -0,0 +1,341 @@
+
+Importing the same Objective-C API in two different ways causes some issues:
+
+* Overloading of synchronous and asynchronous APIs. Objective-C frameworks may have evolved to include both synchronous and asynchronous versions of the same API, e.g.,
+
+  ```objc
+  - (NSString *)lookupName;
+  - (void)lookupNameWithCompletionHandler:(void (^)(NSString *))completion;
+  ```
+  which will be translated into three different Swift methods:
+  
+  ```swift
+  @objc func lookupName() -> String
+  @objc func lookupName(withCompletionHandler: @escaping (String) -> Void)
+  @objc func lookupName() async -> String
+  ```
+  
+  The first and third signatures are identical except for being synchronous and asynchronous, respectively. The async/await design doesn't allow such overloading to be written in the same Swift module, but it can happen when translating Objective-C APIs or when importing methods from different Swift modules. The async/await design accounts for such overloading by favoring synchronous functions in synchronous contexts and asynchronous functions in asynchronous contexts. This overloading should avoid breaking source compatibility.
+
+* Another issue is when an asynchronous completion-handler method is part of an Objective-C protocol. For example, the [`NSURLSessionDataDelegate` protocol](https://developer.apple.com/documentation/foundation/nsurlsessiondatadelegate?language=objc) includes this protocol requirement:
+
+  ```objc
+  @optional
+  - (void)URLSession:(NSURLSession *)session
+            dataTask:(NSURLSessionDataTask *)dataTask
+  didReceiveResponse:(NSURLResponse *)response
+   completionHandler:(void (^)(NSURLSessionResponseDisposition disposition))completionHandler;
+  ```
+
+  Existing Swift code might implement this requirement in a conforming type using its completion-handler signature
+
+  ```swift
+  @objc
+  func urlSession(
+      _ session: URLSession,
+      dataTask: URLSessionDataTask,
+      didReceive response: URLResponse,
+      completionHandler: @escaping (URLSession.ResponseDisposition) -> Void
+  ) { ... }
+  ```
+
+  while Swift code designed to take advantage of the concurrency model would implement this requirement in a conforming type using its `async` signature
+
+  ```swift
+  @objc
+  func urlSession(
+      _ session: URLSession,
+      dataTask: URLSessionDataTask,
+      didReceive response: URLResponse
+  ) async -> URLSession.ResponseDisposition { ... }
+  ```
+
+  Implementing both requirements would produce an error (due to two Swift methods having the same selector), but under the normal Swift rules implementing only one of the requirements will also produce an error (because the other requirement is unsatisfied). Swift’s checking of protocol conformances will be extended to handle the case where multiple (imported) requirements have the same Objective-C selector: in that case, only one of them will be required to be implemented.
```

Comment:
Should this paragraph be changed, to account for the `@optional` requirement?

And/or should a different example with `@required` be used?